--- sqlite-autoconf-3210000/sqlite3.c	2017-10-24 12:26:03.000000000 -0700
+++ sqlite-autoconf-3210000.new/sqlite3.c	2017-11-14 15:36:46.427845830 -0800
@@ -92503,7 +92503,7 @@
 ** Also propagate all EP_Propagate flags from the Expr.x.pList into
 ** Expr.flags. 
 */
-SQLITE_PRIVATE void sqlite3ExprSetHeightAndFlags(Parse *pParse, Expr *p){
+SQLITE_PRIVATE void __attribute__((optimize("O2")))sqlite3ExprSetHeightAndFlags(Parse *pParse, Expr *p){
   if( pParse->nErr ) return;
   exprSetHeight(p);
   sqlite3ExprCheckHeight(pParse, p->nHeight);
